symengine/series_generic.cpp

```diff
@@ -2,155 +2,242 @@
 #include <algorithm>
 #include <iterator>
 #include <symengine/series_generic.h>
-#include <symengine/dict.h>
+#include <symengine/series_visitor.h>
 
 using SymEngine::RCP;
 using SymEngine::make_rcp;
 
 namespace SymEngine {
 
-UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const RCP<const UnivariateIntPolynomial> &poly) :
-        var_{var}, poly_{std::move(poly)} , prec_{precision} {
+RCP<const UnivariateSeries> UnivariateSeries::series(const RCP<const Basic> &t, const std::string &x, unsigned int prec) {
+    UnivariateExprPolynomial p(UnivariatePolynomial::create(symbol(x), {}));
+    SeriesVisitor<UnivariateExprPolynomial, Expression, UnivariateSeries> visitor(std::move(p), x, prec);
+    return visitor.series(t);
 }
 
-UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const unsigned int &max, map_uint_mpz&& dict) :
-        var_{var}, prec_{precision} {
-
-    poly_ = univariate_int_polynomial(var_, std::move(dict));
+std::size_t UnivariateSeries::__hash__() const {
+    return p_.__hash__() + std::size_t(get_degree() * 84728863L);
 }
 
-UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const map_uint_mpz& dict) :
-        var_{var}, prec_{precision} {
+int UnivariateSeries::compare(const Basic &other) const {
+    if (not is_a<UnivariateSeries>(other))
+        throw std::domain_error("cannot compare with UnivariateSeries");
```

Comment:
This should use `SYMENGINE_ASSERT`. See other classes for how this is implemented.